### PR TITLE
ARROW-14014: [Java] Fix Flight parseTrailers for :status keys

### PR DESCRIPTION
**Pull request by BryanCutler**
In Flight Java, the parseTrailers can sometimes receive Metadata with keys that start with ":", such as ":status". Attempting to access the metadata with this type of key causes an error because gRPC-Java rejects these keys as having an invalid prefix. This fixes that by using `InternalMetadata` to create the Key, which checks if the key starts with a ":" and internally labels it as "psuedo" and accepts it as valid.

### CODE REVIEW

**Comment by BryanCutler** [comment]
@lidavidm , I've run into this problem several times but can't figure out how to reproduce it consistently. I used an actual header that was generated for the test, so I believe the fix should work.  I don't know grpc too well, so let me know what you think. Also, there are other places that create Metadata keys, but this is the only one that seemed to cause a problem.

**Comment by lidavidm** [comment]
Do we actually want these headers? How about skipping pseudo-headers?

**Comment by BryanCutler** [comment]
I'm not too sure what types of metadata could be in here, but if they will end up in an error message wouldn't it be good to show the error code from ":status"?

**Comment by lidavidm** [comment]
Well, pseudo-headers are basically just status and authority (effectively "Host"). Status may be nice to have, but gRPC generally doesn't use a meaningful status anyways (Was adding that a recent change? Last I checked, errors were encoded as status 200, with the real status in a normal header.) I suppose it would be useful if you accidentally talk to a non-gRPC endpoint. (I'm still surprised by this; do you have a stack trace or anything, even if not reproducible?)

**Comment by lidavidm** [comment]
Actually, given that reproduction, I wonder if the only time it happens is when talking to a non-gRPC endpoint (or maybe an endpoint that normally proxies a gRPC service properly, but sometimes returns an invalid response from gRPC's perspective?)

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/grpc/StatusUtils.java [hunk 1]
```diff
@@ -121,6 +122,17 @@ public static FlightStatusCode fromGrpcStatusCode(Status.Code code) {
     }
   }
 
+  /** Create Metadata Key for binary metadata. */
+  public static Metadata.Key<byte[]> keyOfBinary(String name) {
```
Comment (lidavidm) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/grpc/StatusUtils.java:126]: Does this need to be public?

**Review by lidavidm — APPROVED** [review]
That said I don't think it hurts to just handle this case.

**Comment by BryanCutler** [comment]
The stack trace from the ticket is below, the author says it happens occasionally due to an intermediate proxy
```
java.lang.IllegalArgumentException: Invalid character ':' in key name ':status'
at com.google.common.base.Preconditions.checkArgument(Preconditions.java:275)
at io.grpc.Metadata$Key.validateName(Metadata.java:742)
at io.grpc.Metadata$Key.<init>(Metadata.java:750)
at io.grpc.Metadata$Key.<init>(Metadata.java:668)
at io.grpc.Metadata$AsciiKey.<init>(Metadata.java:959)
at io.grpc.Metadata$AsciiKey.<init>(Metadata.java:954)
at io.grpc.Metadata$Key.of(Metadata.java:705)
at io.grpc.Metadata$Key.of(Metadata.java:701)
at org.apache.arrow.flight.grpc.StatusUtils.parseTrailers(StatusUtils.java:164)
at org.apache.arrow.flight.grpc.StatusUtils.fromGrpcStatusAndTrailers(StatusUtils.java:128)
at org.apache.arrow.flight.grpc.StatusUtils.fromGrpcRuntimeException(StatusUtils.java:152)
at org.apache.arrow.flight.grpc.StatusUtils.fromThrowable(StatusUtils.java:176)
at org.apache.arrow.flight.FlightClient$SetStreamObserver.onError(FlightClient.java:440)
at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:478)
at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
at org.apache.arrow.flight.grpc.ClientInterceptorAdapter$FlightClientCallListener.onClose(ClientInterceptorAdapter.java:117)
at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:553)
at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:68)
at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:739)
at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:718)
at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:823)
```

We have similar reports, mostly likely due to a proxy and looks like the channel is being closed due to a problem. It's happened to me locally too, and I thought it was during a failed authentication but I haven't been able to reproduce that.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/grpc/StatusUtils.java (hunk 1 above)
Comment (BryanCutler) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/grpc/StatusUtils.java:126]: No, good catch. I was going to update it in other places, but it didn't seem necessary. I'll change to private.

**Comment by lidavidm** [comment]
Thanks. Let's just handle them then, but I guess implementations shouldn't rely on them being present. It's a bit dated, but this ML discussion says gRPC normally removes them; I can't find it from a quick glance, but perhaps (mistakenly or intentionally) they get let through in some cases: https://groups.google.com/g/grpc-io/c/1RPka9CKAvE

**Review by lidavidm — APPROVED** [review]

**Comment by BryanCutler** [comment]
merged to master, thanks @lidavidm !

**Comment by lidavidm** [comment]
Thanks for the fix!

**Pull request closed without merging**